src/cdomains/apron/sharedFunctions.apron.ml

```diff
@@ -112,44 +146,164 @@ struct
             | BinOp (Mod, e1, e2, _) ->
               Binop (Mod, texpr1_expr_of_cil_exp e1, texpr1_expr_of_cil_exp e2, Int, Near)
             | CastE (TInt (t_ik, _) as t, e) ->
-              begin match IntDomain.should_ignore_overflow t_ik || IntDomain.Size.is_cast_injective ~from_type:(Cilfacade.typeOf e) ~to_type:t with (* TODO: unnecessary cast check due to overflow check below? or maybe useful in general to also assume type bounds based on argument types? *)
-                | true ->
-                  Unop (Cast, texpr1_expr_of_cil_exp e, Int, Zero) (* TODO: what does Apron Cast actually do? just for floating point and rounding? *)
```

Comment:
Because the cast analysis was seperated from the overflow analysis and includes the `should_ignore_overflow`. The respective call to handle this is in line 158. If an overflow happens we throw the exception of `Unsupported CilExp`.

Reply:
@DrMichaelPetter and I have considered this again now. We think that having this check before was actually wrong:

Imagine having an `int` and casting it to `short`. The behavior of GCC (and MSVC, Intel, Clang), which is also the one we imitate here says

>> The result of, or the signal raised by, converting an integer to a signed integer type when the value cannot be represented in an object of that type (C90 6.2.1.2, C99 and C11 6.3.1.3).

>For conversion to a type of width N, the value is reduced modulo 2^N to be within range of the type; no signal is raised.

This means that, even when we assure Goblint that no signed overflows occur, it would be incorrect to assume that no cast has happened. Thus `should_ignore_overflow` should not be consulted here.

Does this make sense?